### PR TITLE
Support Go-JSON

### PR DESCRIPTION
**Pull request by zeripath**
This PR provides a mechanism to select the JSON library at start-up, on configuration and at compile time.
Go-JSON support is also added.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
Well that's very interesting - the whole thing breaks!

@silverwind you might wanna take a look

**Comment by a1012112796** [comment]
I think this is ``refactor`` rather than ``feature``

**Comment by silverwind** [comment]
Hmm, I hope we did not hit a bug in that module. I'm certainly worried it might now be as stable as the other alternatives.

**Comment by silverwind** [comment]
Probably need to wait on next version of `go-json` which may have some of these crashes fixed (https://github.com/goccy/go-json/commit/4167aefe08213818e5de6914dd6a9d9bea626bf6).

**Comment by silverwind** [comment]
[0.4.8 is out](https://github.com/goccy/go-json/releases/tag/v0.4.8), please try an update.

**Comment by silverwind** [comment]
Still a crash in there along with some odd test failures.

**Comment by zeripath** [comment]
nope - still crashes.

**Comment by silverwind** [comment]
https://github.com/goccy/go-json/issues/169

**Comment by silverwind** [comment]
Can you try adding [this debug option](https://github.com/goccy/go-json/issues/169#issuecomment-811597199) to the code [here](https://github.com/go-gitea/gitea/blob/3ec2f45f5c719af98b1c1dfc690405e4bd15944c/modules/context/context.go#L373)? I guess that would be easier than me trying to reproduce what the test suite does locally.

**Comment by 6543** [comment]
while running #15250 with master and this head branch ... I did not really notice any difference

**Comment by silverwind** [comment]
Yeah it's likely that those performance claims are exaggerated, or that we just don't do enough JSON encoding that it makes a significant difference.

**Comment by silverwind** [comment]
[0.4.10](https://github.com/goccy/go-json/releases/tag/v0.4.10) is out with a fix for our crash, please update once more.

**Comment by zeripath** [comment]
@silverwind I've given you collaborator access to my fork - so feel free to update

**Comment by silverwind** [comment]
Still another bug present, this time in the decoder it seems. Will check out later.

**Comment by zeripath** [comment]
0.5.0 is out so I'll try to update this again at some point

**Comment by silverwind** [comment]
Bumped to 0.5.1

**Comment by silverwind** [comment]
Not sure if I screwed up or this is a module bug:

```
could not import github.com/goccy/go-json (/drone/src/vendor/github.com/goccy/go-json/encode.go:262:21: DebugRun not declared by package vm_escaped
```

File: modules/json/json.go [hunk 1]
```diff
@@ -67,34 +67,32 @@ func (StdJSON) Indent(dst *bytes.Buffer, src []byte, prefix, indent string) erro
...
 }
 
 // NewDecoder implements Interface
-func (j JSONiter) NewDecoder(reader io.Reader) Decoder {
-	return j.API.NewDecoder(reader)
+func (j GoCcyJson) NewDecoder(reader io.Reader) Decoder {
+	return goccy_json.NewDecoder(reader)
 }
 
 // Indent implements Interface, since jsoniter don't support Indent, just use encoding/json's
```
Comment (6543) [inline modules/json/json.go]:
```suggestion
// Indent implements Interface
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -142,14 +142,14 @@ For imports you should use the following format (_without_ the comments)
...
   // stdlib
-  "encoding/json"
   "fmt"
 
   // local packages
   "code.gitea.io/gitea/models"
   "code.gitea.io/sdk/gitea"
 
   // external packages
+  "github.com/goccy/go-json"
```
Comment (techknowlogick) [inline CONTRIBUTING.md]: This should be moved above, and referenced as `"code.gitea.io/gitea/modules/json"`

**Comment by zeripath** [comment]
OK so looking a bit more closely at the first unit-test error

```json
{
  "push_only":true,
  // ... nil values ...
  {
    "push":true,
    // ... more nil values ...
  }
}
``` doesn't seem to get unmarshalled properly in to  ```go
models.HookEvent{PushOnly:true, ..., HookEvents:models.HookEvents{Push:true, ...}}
``` instead becomes: ```go
models.HookEvent{PushOnly:true, ..., HookEvents:models.HookEvents{Push:false, ...}}
```

**Comment by DasSkelett** [comment]
As someone who followed the development of goccy/go-json for a few months, I can not recommend to put it in any sort of production setup as of now. While it is very fast compared to alternatives, it is very much alpha quality in my eyes. I would not want to see it already used in Gitea as Gitea user.
Especially if JSON encoding/decoding is not part of "the critical code path" that somehow limits performance of the whole server, I would not risk it. Maybe give it some time and keep an eye on it, check back when it hits 1.0.0.

**Comment by zeripath** [comment]
@DasSkelett I am in agreement with you on this. Whilst Goccy appears interesting there still appear to be several very important bugs.

https://github.com/goccy/go-json/issues/270

**Comment by silverwind** [comment]
Maybe close this PR and revisit once it's 1.0.

**Comment by zeripath** [comment]
I'm happy to keep updating this tbh. It's a good way to keep an eye on whether this is ready or not.

**Comment by zeripath** [comment]
```
--- FAIL: TestHTTPClientDownload (0.00s)
    http_client_test.go:255: 
        	Error Trace:	http_client_test.go:255
        	Error:      	Should be true
        	Test:       	TestHTTPClientDownload
        	Messages:   	case 1: 'not at beginning of value' should contain 'invalid json'
2021/08/12 11:25:11 ...ules/lfs/endpoint.go:73:endpointFromURL() [E] lfs.endpointFromUrl: unknown url
2021/08/12 11:25:11 ...s/lfs/http_client.go:106:batch() [E] Error decoding json: not at beginning of value
2021/08/12 11:25:11 ...s/lfs/http_client.go:106:batch() [E] Error decoding json: not at beginning of value
--- FAIL: TestHTTPClientUpload (0.00s)
    http_client_test.go:356: 
        	Error Trace:	http_client_test.go:356
        	Error:      	Should be true
        	Test:       	TestHTTPClientUpload
        	Messages:   	case 1: 'not at beginning of value' should contain 'invalid json'
FAIL
```

This one appears to be a difference in the way invalid json is reported.

**Comment by silverwind** [comment]
Yeah, some error messages like that do not seem to match `encodings/json`, I guess we should report as go-json aims to be a drop-in replacement.

**Comment by zeripath** [comment]
OK it finally looks like we're passing tests! 

I think https://github.com/goccy/go-json/issues/275 needs to be fixed and we need to test actually running gitea with this backend but it might actually be getting close.

**Comment by zeripath** [comment]
OK GoCcy now works. 

My suspicion is that we shouldn't run with goccy by default but question is whether we can make the JSON variant selectable.

The problem is that as it currently stands `modules/json.DefaultJSONHandler` is not concurrent safe so we should set choose the JSON variant right at the start of gitea before any other goroutines start that might use it. This likely precludes its use in `setting.NewContext`.

Currently I've provided a way for the choice to be compiled in at link time by overwriting a variable at link time. e.g. `LDFLAGS="-X code.gitea.io/gitea/modules/json.DefaultJSONHandlerType=std"`.

File: modules/json/json.go [hunk 3]
```diff
@@ -31,14 +33,41 @@ type Interface interface {
...
+	switch strings.ToLower(typ) {
+	case "std":
+		DefaultJSONHandler = StdJSON{}
+		DefaultJSONHandlerType = "std"
+	case "goccy":
+		DefaultJSONHandler = GoCcyJSON{}
+		DefaultJSONHandlerType = "goccy"
+	case "jsoniter":
+		DefaultJSONHandler = JSONiter{jsoniter.ConfigCompatibleWithStandardLibrary}
+		DefaultJSONHandlerType = "jsoniter"
```
Comment (lunny) [inline modules/json/json.go]: fallthrough

**Review by techknowlogick — APPROVED** [review]

**Comment by wxiaoguang** [comment]
I also have the question why do we need to make JSON library configurable.

Does it bring more benefits than the complexity?

**Review by lunny — CHANGES_REQUESTED** [review]
Sorry I will against make it configurable for end users. It should be a build tag.

**Comment by wxiaoguang** [comment]
> Sorry I will against make it configurable for end users. It should be a build tag.

I would not even introduce a build tag. If I find something is good enough, just I will use it.

If there is no real benefit, we do not need to introduce too many options/tags.

**Comment by silverwind** [comment]
Just checking the issue tracker of `go-json`, it still has crash bugs. Maybe revisit again in a few years if it ever becomes crash-free. 

I suggest sticking to `encoding/json`, as it will receive the more developer attention than those unstable/abandoned hobby projects.

**Comment by lunny** [comment]
> Just checking the issue tracker of `go-json`, it still has crash bugs.
> 
> I suggest sticking to `encoding/json`, as it will receive much more developer attention than those unstable/abandoned hobby projects and is bound to only get faster over time.

I suggest we can close this one and open a new when something is ready.  @zeripath

**Comment by zeripath** [comment]
When I last updated and fixed this PR go-json was working completely fine for Gitea - but the potential issue of there being bugs was a reason why I provided the configuration variable which some reviewers did not like.

@silverwind we're already using json-iterator over encoding/json and I would note that encoding/json has not increased in speed significantly for the entire period time I have been involved in gitea. 

However, it's clear that this PR is not going to be merged so I will close it.

**Comment by silverwind** [comment]
I recommend dropping `json-iterator` as well. It has a number of mind-bogging bugs. Gitea might not hit them currently, but it can in the future.

**Comment by zeripath** [comment]
There is a *considerable* time penalty to using encoding/json over jsoniterator. You are more than welcome to write the PR and fix the multiple issues were we have relied on jsoniterator error interpretation (this PR has the fixes for those) and then do the performance tests to prove this.

**Comment by silverwind** [comment]
I guess we shouldn't try to fix what aint broken then.

On topic of benchmarks, I do find https://medium.com/geekculture/my-golang-json-evaluation-20a9ca6ef79c from 2021, so I guess you are right, `encodings/json` is worst:

image

**Pull request closed without merging**